compiler/types.nim

```diff
@@ -1471,6 +1494,76 @@ proc skipHiddenSubConv*(n: PNode; g: ModuleGraph; idgen: IdGenerator): PNode =
   else:
     result = n
 
+proc getProcConvMismatch*(c: ConfigRef, f, a: PType, rel: var TTypeRelation): set[ProcConvMismatch] =
+  ## Returns a set of the reason of mismatch.
+  ## `rel` is mutated on fail to `isNone`.
+  
+  if tfNoSideEffect in f.flags and tfNoSideEffect notin a.flags:
+    # Formal is pure, but actual is not
+    result.incl pcmNoSideEffect
+    rel = isNone
+
+  if tfThread in f.flags and a.flags * {tfThread, tfNoSideEffect} == {} and
+    optThreadAnalysis in c.globalOptions:
+    # noSideEffect implies ``tfThread``!
+    result.incl pcmNotGcSafe
+    rel = isNone
+
+  if f.flags * {tfIterator} != a.flags * {tfIterator}:
+    # One of them is an iterator so not convertible
+    result.incl pcmNotIterator
+    rel = isNone
+
+  if f.callConv != a.callConv:
+      # valid to pass a 'nimcall' thingie to 'closure':
+      if f.callConv == ccClosure and a.callConv == ccNimCall:
+        case rel
+        of isInferred: rel = isInferredConvertible
+        of isBothMetaConvertible: rel = isBothMetaConvertible
+        elif rel != isNone: rel = isConvertible
+      else:
+        rel = isNone
+        result.incl pcmDifferentCallConv
+
+  if f.lockLevel.ord != UnspecifiedLockLevel.ord and
+     a.lockLevel.ord != UnspecifiedLockLevel.ord:
+       # proctypeRel has more logic to catch this difference,
+       # so dont need to do `rel = isNone`
+       # but it's a pragma mismatch reason which is why it's here
+       result.incl pcmLockDifference
+
+proc getProcConvMismatch*(c: ConfigRef, f, a: PType): set[ProcConvMismatch] =
+  var rel: TTypeRelation
+  getProcConvMismatch(c, f, a, rel)
+
+
+proc addPragmaAndCallConvMismatch*(message: var string, formal, actual: PType, conf: ConfigRef) =
+  assert formal.kind == tyProc and actual.kind == tyProc
+  let convMismatch = getProcConvMismatch(conf, formal, actual)
+  var
+    gotPragmas = ""
+    expectedPragmas = ""
+  for reason in convMismatch:
+    case reason:
+    of pcmDifferentCallConv:
+      let
+        got = $(actual.callConv)
+        expected = $(formal.callConv)
+      message.add "\n  Calling convention mismatch: got '{.$1.}', but expected '{.$2.}'." % [got, expected]
+    of pcmNoSideEffect:
+      expectedPragmas.add "noSideEffect, "
+    of pcmNotGcSafe:
+      expectedPragmas.add "gcsafe, "
+    of pcmLockDifference:
+      gotPragmas.add("locks: " & $actual.lockLevel & ", ")
+      expectedPragmas.add("locks: " & $formal.lockLevel & ", ")
+    else: discard
```

Comment:
bad style, better be explicit about remaining branches so that if we add a `reason` enum member, we'll get a CT error and be forced to choose what to do with it instead of it silently going to else branch